File: 102 Binary Tree Level Order Traversal/102 Binary Tree Level Order Traversal.py
"""
Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).

Example 1:

Input: root = [3,9,20,null,null,15,7]
Output: [[3],[9,20],[15,7]]

Example 2:

Input: root = [1]
Output: [[1]]

Example 3:

Input: root = []
Output: []
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

        if(not root):   return []
        queue = collections.deque([root])
        all_levels = []
        while (queue):
            size = len(queue)
            curr_level = []
            for _ in range(size):
                curr = queue.popleft()
                curr_level.append(curr.val)
                if(curr.left):     queue.append(curr.left)
                if(curr.right):    queue.append(curr.right)
            all_levels.append(curr_level)
        return all_levels

        ## RECURSIVE APPROACH ##
        if not root:    return []
        levels = collections.defaultdict(list)
        def helper(node, level):
            levels[level].append(node.val)
            if node.left:   helper(node.left, level + 1)
            if node.right:  helper(node.right, level + 1)
        helper(root, 0)
        return levels.values()

File: 102 Binary Tree Level Order Traversal/102 Binary Tree Level Order Traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
    res := [][]int{}
    traversal(root, 1, &res)
    return res
}

func traversal(root *TreeNode, level int, res *[][]int) {
    if root == nil {
        return
    }

    if len(*res) + 1 == level {
        row := []int{}
        *res = append(*res, row)
    }

    (*res)[level-1] = append((*res)[level-1], root.Val)

    traversal(root.Left, level+1, res)
    traversal(root.Right, level+1, res)
}
